### PR TITLE
Disable duplicated flags

### PR DESCRIPTION
**Pull request by zacharya19**
The autocompletion continue to offer flags that already in the command, so I added a check from parsed_args.
Not sure if it's the best way, will be glad to get more ideas on it.

### CODE REVIEW

**Comment by zacharya19** [comment]
Just saw a had a mistake there - fixing it now.

**Comment by kislyuk** [comment]
Thanks for your contribution, @zacharya19. By default, argcomplete stays as close as possible to argparse semantics. This means offering any completions that will result in a successful parse by argparse. Because passing the same optional with different values is allowed by argparse (and desired by some use cases), the behavior that you propose will have to be optional. There are several ways to accomplish this; one way that I would like to try is if you could create a subclass of CompletionFinder and see if you could add your behavior with minimal changes. If you have trouble doing so, I'd like to find out what needs to be changed to make it easy to subclass CompletionFinder in this way.

**Comment by zacharya19** [comment]
@kislyuk Because it's just a little changes on the code, why not pass a new arg to CompletionFinder claas with default False? Why create subclass?
something like that:
```python
if self.no_duplicate:
    completions = [option for option in completions if option[2:] not in parsed_args]
```

On filter_completions function.

**Comment by kislyuk** [comment]
I don't think adding more keyword arguments to the default constructor is a sustainable extensibility strategy. There are already a couple of keyword arguments on the constructor that could probably be best accommodated as extensions. I'm not completely opposed to a keyword argument, but only if it's combined with a modular and easy to read solution.

**Comment by zacharya19** [comment]
Sorry for all the stupid commits :(
Hope it's good, I also found a way to add my other feature request (#144).

**Comment by zacharya19** [comment]
Should I put the _filter_groups function on CompletionFinder and not ExclusiveCompletionFinder? Since it's not possible on argparse to put two args in the same group.

**Comment by kislyuk** [comment]
@zacharya19, thanks, I'm still working on a review for this. In the meantime, please take a look at the changes introduced by @evanunderscore in #145, and adjust this if needed. I chose that solution for the mutex groups part of the problem, because it's more concise, easier to follow, and leverages logic already built into argparse.

**Comment by evanunderscore** [comment]
It looks like your implementation will still suggest an action that has already been specified if the value it stored happened to be falsy. You're also going to run into problems for actions that share destination names. In general any kind of solution involving `parsed_args` is likely going to have these kinds of problems.

It would be better if you could find a cleaner way of working out which actions have already been specified. You might want to take a look at what I did for #145 - I haven't tested this, but doing something similar with `seen_actions` looks like a good idea.

**Comment by zacharya19** [comment]
Ok, I think now it's better, but still could not find a solution for args with "append" action.
Any idea?

**Comment by evanunderscore** [comment]
That looks great. I originally thought this should use `seen_actions` but I think you're correct to use `seen_non_default_actions`. There may be some weirdness as described in http://bugs.python.org/issue18943 but I don't think dealing with that should be our problem.

For "append" actions, the action itself should be an `argparse._AppendAction` instance. You should also take "append_const" into account.

**Comment by zacharya19** [comment]
Thanks! Done!
Look like it's working for append_const as well.

**Comment by evanunderscore** [comment]
@kislyuk What's your feeling on subclass versus argument? Given how simple the solution looks I'd lean towards argument, but it's your call.

@zacharya19 I can't find a reference to an `_orig_class` attribute anywhere in argparse for 3.5 or 2.7. Are you sure that works? Would a simple `isinstance` check be better?

You will also have to add some tests to `test/test.py` to verify the changes you've made. Try to follow the pattern used in the rest of the file - again it might help to look at #145. Make sure you exercise (at least) the regular case, the append case, and the append_const case. If this is to be implemented as a subclass you'll have to make some small changes to the `run_completer` helper.

**Comment by kislyuk** [comment]
@evanunderscore, `_orig_class` is set by argcomplete: https://github.com/kislyuk/argcomplete/blob/master/argcomplete/__init__.py#L290

I am not opposed to a kwarg based solution if it's small and readable, and orthogonal to other filtering logic.

You are right that there should be a test before this can be merged.

**Comment by evanunderscore** [comment]
Oh, I totally missed that. That makes sense.

**Comment by zacharya19** [comment]
Any update here?

**Comment by evanunderscore** [comment]
I can give you my thoughts, but ultimately it's up to @kislyuk 

* The way you're calling `super` is incorrect (but happens to work in this case). If it makes the code easier to reason about, you can just take the `staticmethod` decorator off the base class and add self as a parameter.
* The test case is in the wrong class - it belongs under `TestArgcomplete` rather than `TestArgcompleteREPL`. Moving it might take a little bit of effort to make it work in the same way.
* The test case covers `append_const` twice and doesn't use the plain `append`. I would change one of those.

I also still think this would work better as an extra argument than as a subclass. If you have time, perhaps you could (on a separate branch) modify this to work that way so we can compare the two side by side?

File: argcomplete/__init__.py [hunk 1]
```diff
@@ -600,6 +600,17 @@ def display_completions(substitution, matches, longest_match_length):
         """
         return self._display_completions
 
+class ExclusiveCompletionFinder(CompletionFinder):
+    @staticmethod
+    def _action_allowed(action, parser):
+        class_name = ExclusiveCompletionFinder
+        allowed = super(class_name, class_name)._action_allowed(action, parser)
```
Comment (kislyuk) [inline argcomplete/__init__.py]: Please fix the use of super as mentioned by @evanunderscore.

File: argcomplete/__init__.py [hunk 2]
```diff
@@ -600,6 +600,17 @@ def display_completions(substitution, matches, longest_match_length):
...
+
+        append_classes = (argparse._AppendAction, argparse._AppendConstAction)
+        not_exist = action not in parser._seen_non_default_actions or action._orig_class in append_classes
```
Comment (kislyuk) [inline argcomplete/__init__.py]: Please rewrite this line. The variable name `not_exist` is unclear, please rename it. Please break up the logical statement into a more readable form.

**Comment by zacharya19** [comment]
Not sure if I broke up the logical statement to a better form, what do you think?

File: test/test.py [hunk 3]
```diff
@@ -612,6 +614,26 @@ def make_parser():
...
+        parser.add_argument("--baz", choices=["baz1", "baz2"])
+        parser.add_argument("--no-bar", action="store_true")
+
+        completer = ExclusiveCompletionFinder(parser, always_complete_options=True)
+
+        expected_outputs = (
+            ("prog ", ["--foo", "--bar", "--baz", "--no-bar"]),
+            ("prog --baz ", ["baz1", "baz2"]),
+            ("prog --baz baz1 ", ["--foo", "--bar", "--no-bar"]),
+            ("prog --bar ", ["bar1", "bar2"]),
```
Comment (evanunderscore) [inline test/test.py]: This test case is a duplicate of the one 2 lines above.

**Comment by evanunderscore** [comment]
That's looking better. You'll need to resolve the merge conflicts and check that the CI is passing too.

**Comment by zacharya19** [comment]
Something want wrong with the test, not sure what.
Any help?

**Comment by evanunderscore** [comment]
Click "Details", then click one of the failed builds to see the log. This is the relevant part:

```
$ make lint

./setup.py flake8

running flake8

argcomplete/__init__.py:615:1: W293 blank line contains whitespace

make: *** [lint] Error 1
```

**Comment by zacharya19** [comment]
I was talking about this:
```
.............................usage: test.py [-h] --foo FOO {bar}
test.py: error: too few arguments
usage: test.py [-h] --foo FOO {bar}
test.py: error: argument --foo is required
...usage: test.py foo [-h] {bar}
test.py foo: error: too few arguments
...x.x......x......x.x......x.....x.xEEEEEEEEEEEEEEEE
======================================================================
ERROR: test_continuation (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_double_quoted_completion (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_exclamation_in_double_quotes (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_parse_special_characters (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_parse_special_characters_dollar (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_partial_completion (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_quoted_exact (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_quoted_space (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_simple_completion (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_single_quoted_completion (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_single_quotes_in_double_quotes (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_single_quotes_in_single_quotes (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_special_characters (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_special_characters_double_quoted (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_special_characters_single_quoted (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

======================================================================
ERROR: test_unquoted_space (__main__.TestTcsh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 934, in setUp
    sh = Shell('tcsh')
  File "test.py", line 944, in __init__
    self.child = pexpect.spawn(shell, encoding='utf-8')
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 194, in __init__
    self._spawn(command, args, preexec_fn, dimensions)
  File "/usr/lib/python2.7/dist-packages/pexpect/pty_spawn.py", line 267, in _spawn
    'executable: %s.' % self.command)
ExceptionPexpect: The command was not found or was not executable: tcsh.

----------------------------------------------------------------------
Ran 85 tests in 36.240s

FAILED (errors=16, expected failures=8)
```

**Comment by kislyuk** [comment]
Great work, thanks. I would also suggest that you add a docstring (you can open a separate PR) so this shows up in the docs.

**Pull request merged**